tests/src/core/testqgsrulebasedrenderer.cpp

```diff
@@ -1152,6 +1152,38 @@ class TestQgsRuleBasedRenderer: public QgsTest
       QCOMPARE( counter->featureCount( "2" ), 1LL );
     }
 
+    void testNullsCountForString()
```

Comment:
This test should be in test_qgscategorizedsymbolrenderer.py, not amongst the rule based renderer tests. Are you happy to port to Python and move it there?

Reply:
Yes, very happy to write it in Python. Sadly, I still couldn't write a failing test as for some reason the feature counts are correct when counting feature with QgsVectorLayerFeatureCounter. I also noticed while testing manually in QGIS that the issue only happens if "Show Feature Counts" checkbox is checked before the categories are added. So could it be related to some signal not firing correctly? When the checkbox is checked after Classify button is pressed, the counts seem to be correct. I tried to replicate the behaviour by running the following code in python console but the counts are still correct.

```python
# create layer
layer = QgsVectorLayer("Point?field=test_field:string", "test_layer", "memory")
fields = QgsFields()
fields.append(QgsField('test_field', QVariant.String))
for attr_value in ['a'] * 3 + ['b'] * 2 + [None]:
    f = QgsFeature(fields)
    f.setAttributes([attr_value])
    layer.dataProvider().addFeature(f)

# add layer to map and show counts
QgsProject.instance().addMapLayer(layer)
root = QgsProject.instance().layerTreeRoot()
layer_node = root.findLayer(layer.id())
layer_node.setCustomProperty("showFeatureCount", True)

renderer = QgsCategorizedSymbolRenderer()
renderer.setClassAttribute('test_field')
renderer.addCategory(QgsRendererCategory('a', QgsMarkerSymbol(), 'a'))
renderer.addCategory(QgsRendererCategory('b', QgsMarkerSymbol(), 'b'))
renderer.addCategory(QgsRendererCategory(None, QgsMarkerSymbol(), 'nulls'))
layer.setRenderer(renderer)

# this shows correct count 1
layer.featureCount('2')
```